src/plots/cartesian/axes.js

```diff
@@ -657,8 +657,35 @@ function isClose(a, b) {
 axes.prepTicks = function(ax, opts) {
     var rng = Lib.simpleMap(ax.range, ax.r2l, undefined, undefined, opts);
 
-    // calculate max number of (auto) ticks to display based on plot size
-    if(ax.tickmode === 'auto' || !ax.dtick) {
+    // sync ticks with the overlaying defined axis
+    if(ax.tickmode === 'sync') {
+        var baseAxis = ax._mainAxis;
+
+        // get range min and max to find range delta of axis 1
+        var minValBaseAxis = Math.min(baseAxis.range[0], baseAxis.range[1]);
+        var maxValBaseAxis = Math.max(baseAxis.range[0], baseAxis.range[1]);
+        var rangeDeltaBaseAxis = maxValBaseAxis - minValBaseAxis;
+
+        // get range min and max to find range delta of axis 2
+        var minValAxis = Math.min(ax.range[0], ax.range[1]);
+        var maxValAxis = Math.max(ax.range[0], ax.range[1]);
+        var rangeDeltaCurrentAxis = maxValAxis - minValAxis;
+
+        // set second axis' dtick value to be based off of same ratio as the first axis
+        var dtickRatio = rangeDeltaBaseAxis / baseAxis.dtick;
+        ax.dtick = rangeDeltaCurrentAxis / dtickRatio;
```

Comment:
I think we need to make use of `Lib.increment` here.
Otherwise we would encounter rounding errors.
For example if I call 
```js
Plotly.relayout(gd, 'yaxis2.range', [0, 0.5]);
```
On your mock at the moment, this is what I get:
 
![z-new_tickmode_sync-relayout](https://user-images.githubusercontent.com/33888540/202542784-002e217f-240e-4b08-a971-721ad09f93fe.png)

Reply:
On the other hand, instead of using `ax.range` which include offsets on each side, you should use e.g. first and last grid values i.e. `axis.tick0` and axis._gridVals[axis._gridVals.length - 1].x` in your calculation above.

Reply:
I was trying to address your comment but it's not super clear to me. If I use the Lib.Increment both grid lines don't align perfectly. I don't know what to do in order to not have rounding issues and the lines stay aligned.

About your second comment, using tick0 and last grid value doesn't seem to be clear to me especially if the overlaying axis is tickmode = auto so tick0 = 0 . Could you please give me more details?

All others comments are addressed in my latest commit.

Thanks

Reply:
> @archmoj I was trying to address your comment but it's not super clear to me. If I use the Lib.Increment both grid lines don't align perfectly. I don't know what to do in order to not have rounding issues and the lines stay aligned.
> 

Basically  [`Lib.increment`](https://github.com/plotly/plotly.js/blob/master/src/lib/increment.js) helps avoid certain floating errors, as illustrated in this [test file](https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/lib_increment_test.js).

Did you manage to use it?
If so could you please share an image? 
If grid lines has 1 pixel difference then perhaps we could skip drawing the second grid over. 
In fact I think that's what we should do in general to handle semi transparent grid lines.

Reply:
@archmoj I'm using like that:
![image](https://user-images.githubusercontent.com/33151376/204314200-671f7224-51b4-4e35-a178-a9de49f71616.png)

Is this the correct way?

Reply:
@archmoj I'm using like that:
![image](https://user-images.githubusercontent.com/117312714/204314665-7d0dbd10-558f-4c27-8ccd-d38842fd9479.png)

Is this the correct way?

Reply:
Thanks for the screenshot.
The problem may not be with `tick0` itself but how we add the offset to compute next values.
BTW you could commit your workaround and we could test that.

Reply:
I don't think it's a rounding problem but when you ask for `Plotly.relayout(gd, 'yaxis2.range', [0, 0.5]);` it subtracts 1 pixel from the top and 1 pixel from the bottom. If I call `Plotly.relayout(gd, 'yaxis2.range', [0, 0.501]);` eveything will be ok. 

I don't have a proper workaround at the minute and to be honest I'm a bit lost what's the direction I should go to fix this issue.

Reply:
> @archmoj I don't think it's a rounding problem but when you ask for `Plotly.relayout(gd, 'yaxis2.range', [0, 0.5]);` it subtracts 1 pixel from the top and 1 pixel from the bottom. If I call `Plotly.relayout(gd, 'yaxis2.range', [0, 0.501]);` eveything will be ok.
> 
> I don't have a proper workaround at the minute and to be honest I'm a bit lost what's the direction I should go to fix this issue.

OK. After 
```js
Plotly.relayout(gd, 'yaxis2.range', [0, 0.5]);
```
`gd._fullLayout.yaxis2.dtick` is set to 0.09976057462090983 but that should be `0.1` or very close to `0.1`.
Perhaps the computation for `ax.dtick` is not quite accurate?

Reply:
Could you please try something like:
```js
    if(ax.tickmode === 'sync') {
        var baseAxis = ax._mainAxis;

        // get range min and max to find range delta of axis 2
        var minValAxis = Math.min(ax.range[0], ax.range[1]);
        var maxValAxis = Math.max(ax.range[0], ax.range[1]);
        var rangeDeltaCurrentAxis = maxValAxis - minValAxis;

        // validate that the axis has values before we try to use them for tick0 calculation
        if(baseAxis && baseAxis._vals.length > 0) {
            var _len = baseAxis._length - 1;

            // get position of first & last axis' starting ticks
            var firstTickPosition = baseAxis.l2p(baseAxis._vals[0].x);
            var firstTickPercentage = firstTickPosition / _len;
            firstTickPercentage = Math.round(firstTickPercentage * 100) / 100;

            var lastTickPosition = baseAxis.l2p(baseAxis._vals[baseAxis._vals.length - 1].x);
            var lastTickPercentage = lastTickPosition / _len;
            lastTickPercentage = Math.round(lastTickPercentage * 100) / 100;

            // set current axis to have same starting tick position
            var firstOffset = rangeDeltaCurrentAxis * (1 - firstTickPercentage);
            ax.tick0 = firstOffset + minValAxis;
            
            var lastOffset = rangeDeltaCurrentAxis * (1 - lastTickPercentage);
            ax.dtick = (lastOffset - firstOffset) / (baseAxis._vals.length - 1);
        }
    } else if ...
```

Reply:
@archmoj I've changed the code to your suggestion but some lines don't align perfectly. 

![image](https://user-images.githubusercontent.com/117312714/205007774-9743711d-50d6-4433-839a-1b1394b95c8e.png)

Would be perfect to hide one of the lines but if and only if we can sync both axes to move together. At the moment only the overlayed axis can move together, not the main one.